solr/solrj/src/java/org/apache/solr/client/solrj/cloud/autoscaling/Policy.java

```diff
@@ -644,15 +643,13 @@ public Session cloneToNewSession(SolrCloudManager cloudManager) {
       List<Clause> expandedClauses = new ArrayList<>(this.expandedClauses);
 
       List<Row> matrix = new ArrayList<>(nodes.size());
+      Map<String, Row> copyNodes = new HashMap<>();
+      for (Row oldRow: this.matrix) {
```

Comment:
Tradeoff with this one time matrix walk and copy is that we copy all nodes, even those that might not be needed (because not present in parameter `nodes`). I do like the approach though it's cleaner than the code I wrote.

Maybe keep this initial loop but just store Row references, then do the copy below when newRow is needed? (there's already an if bloc when it's null, so copy action can easily be placed in else bloc).

Reply:
>even those that might not be needed

I believe it's a reasonable tradeoff because most of the times nodes do not go away. Yes, we make an extra copy in those rare occasions, but it should be OK.